### ISSUE TITLE
None type inference is incorrect for option type when None is checked against a list

### ISSUE THREAD

**Issue by shassard** [microsoft/pyright#1714]
**Describe the bug**

Checking for None in a list of values and returning early doesn't clear None from a list of possible values.

**To Reproduce**

In the code below, the type checker incorrectly believes that None is a possible value for *some_input* in the function bad_none_type():

**Expected behavior**

The type inferred bad_none_type() Optional arguments should be non-None values.

**Screenshots or Code**

```python
from typing import Optional

def bad_none_type(some_input: Optional[str], other_input: Optional[str]):
    if None in [some_input, other_input]:
        return
    # incorrect type below: "capitalize" is not a known member of "None"
    print("bad", some_input.capitalize(), other_input.capitalize())

def good_none_type(some_input: Optional[str], other_input: Optional[str]):
    if some_input is None or other_input is None:
        return
    print("good", some_input.capitalize(), other_input.capitalize())

def sanity_check():
    good_none_type("foo", "bar")
    good_none_type(None, "bar")
    good_none_type("foo", None)
    good_none_type(None, None)

    bad_none_type("foo", "bar")
    bad_none_type(None, "bar")
    bad_none_type("foo", None)
    bad_none_type(None, None)

if __name__ == "__main__":
    sanity_check()
```

The output of the above script in python 3.9.2 is:
```
PS C:\Users\steve> & C:/Users/steve/AppData/Local/Programs/Python/Python39/python.exe d:/Devel/pyright_none_bug.py
good Foo Bar
bad Foo Bar      
```

**VS Code extension or command-line**

Using pylance 2021.3.4 in vscode 1.55.0

**Comment by erictraut** [comment]
This isn't one of the supported type narrowing patterns supported by pyright. For a complete list of supported patterns, refer to [this documentation](https://github.com/microsoft/pyright/blob/master/docs/type-concepts.md#type-guards).

Each supported type narrowing pattern requires special-case logic and adds complexity (and in some cases potential perf degradation) to the type checker, so we support only common patterns. To my knowledge, pyright already supports more type narrowing patterns than any other Python type checker. I don't think this pattern is common enough to justify adding there's an easy workaround in this case — one that is faster at runtime and makes for more readable code:

```python
if some_input is None or other_input is None:
    return
```
